js/app.js

```diff
@@ -84,18 +85,42 @@ const App = {
 			}
 		}
 	},
+	async login() {
+		// Login removes the document fragment, so keep it in storage
+		// See https://www.rfc-editor.org/rfc/rfc6749#section-3.1.2
+		if (getAppLocation() != null) {
+			sessionStorage.setItem("docLocation", document.location.hash);
+			setWindowURLFragment(null);
+		}
+		const keycloak = new Keycloak({
+			url: "https://app.please-open.it",
+			realm: "5d970145-a48d-4e9b-84ab-15de19f9d3a5",
+			clientId: window.location.hostname
+		});
+		await keycloak.init({
+			onLoad: "login-required",
+			flow: "implicit",
+			checkLoginIframe: false
+		});
+		let docLocation = sessionStorage.getItem("docLocation");
+		if (docLocation != null) {
+			sessionStorage.removeItem("docLocation");
+			setWindowURLFragment(docLocation);
+		}
+		return keycloak.token;
+	},
 	async init() {
 		function onHashChange() {
 			let {documentId, filter} = getAppLocation() ?? {};
-			const isNew = !documentId;
-			if (isNew) {
-				documentId = uuid();
+			if (!documentId) {
+				documentId = (documentId ?? localStorage.getItem("lastTodoList")) || uuid();
 				setAppLocation(documentId, filter);
 			}
+			localStorage.setItem("lastTodoList", documentId);
```

Comment:
Here we change the behaviour of the app to redirect itself back to the user's last TodoList if no URL fragment was specified. This is possible because the use of [Named Subdomains](https://gw.m-ld.org/named-subdomains) in the Gateway means the TodoLists are stored there. This is a significant upgrade to the application: a user can navigate to their stored TodoLists any time on any device, without having to have them open elsewhere.